# Segmenting and Clustering Neighborhoods in Toronto
Created by Scott Proulx
Last edited: May 29, 2020

## Purpose of this notebook:
The purpose of this notebook is to house code for the course *Applied Data Science Capstone* on Coursera, as part of the *IBM Data Science Professional Certificate*. In particular, the first graded assignment will be shown here.

First, let's import some of the packages we will be using:

In [216]:
import pandas as pd
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)
import numpy as np

## Part 1 - Creating Toronto DataFrame

In this section, we will be scraping Wikipedia for neighborhood data. First, let's import Beautiful Soup:

In [217]:
!conda install -c conda-forge beautifulsoup4 --yes 
print("Beautiful Soup installed successfully!")

Solving environment: ...working... 
  - anaconda/win-64::openssl-1.1.1d-he774522_2
  - defaults/win-64::openssl-1.1.1d-he774522_2done

# All requested packages already installed.

Beautiful Soup installed successfully!


Import the necessary modules:

In [218]:
from bs4 import BeautifulSoup
import requests

Proceed to scraping the URL:

In [219]:
# Enter the URL to visit:
URL = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'

# Find the table on the page:
res = requests.get(URL).text
soup = BeautifulSoup(res,'lxml')
my_table = soup.find('table', {'class':'wikitable sortable'})

# Initialize a list for each column:
postal_code = []
borough = []
neighborhood = []

# Go through each row of the table.
for items in my_table.find_all("tr")[1:-1]:
    data = [' '.join(item.text.split()) for item in items.find_all(['th', 'td'])]
    postal_code.append(data[0])
    borough.append(data[1])
    neighborhood.append(data[2])

Now we will build a Pandas Dataframe with the table data:

In [220]:
df_toronto = pd.DataFrame({'Postal Code':postal_code, 'Borough':borough, 'Neighborhood':neighborhood})
df_toronto.head()

Postal Code           Borough               Neighborhood
0         M1A      Not assigned                           
1         M2A      Not assigned                           
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront

Now we will remove all rows that have an unassigned borough:

In [221]:
df_toronto = df_toronto[df_toronto['Borough'] != 'Not assigned']
df_toronto.reset_index(drop = True, inplace = True)
df_toronto.head()

Postal Code           Borough                                 Neighborhood
0         M3A        North York                                    Parkwoods
1         M4A        North York                             Victoria Village
2         M5A  Downtown Toronto                    Regent Park, Harbourfront
3         M6A        North York             Lawrence Manor, Lawrence Heights
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

Now let's check to see if there are any duplicates in the Postal Code section:

In [222]:
print("Number of postal codes:", len(df_toronto['Postal Code']))
print("Number of unique postal codes:", len(df_toronto['Postal Code'].unique()))

Number of postal codes: 103
Number of unique postal codes: 103


Therefore, it appears that all the postal codes are already unique and do not need to be handled.
Now, let us check for instances where a Borough has been assigned but a Neighborhood has NOT been assigned (note that we have already removed all instances where a Borough was not assigned):

In [223]:
print("Number of unassigned Neighborhoods:", len(df_toronto[df_toronto['Neighborhood'] == '']))

Number of unassigned Neighborhoods: 0


Therefore, we do not need to modify the data to assign neighborhood names.

Shown below is the size of our Toronto neighborhood dataframe:

In [224]:
df_toronto.shape

(103, 3)

Therefore, there are 103 different postal codes in Toronto.

## Part 2 - Determining Coordinates of Postal Codes

In this section, we are going to find the coordinates for each postal code obtained in the above dataframe. Note, that I tried to use Geocoder, but even after thousands of calls I was unable to obtain coordinates for several postal codes. Therefore, we will be obtaining the coordinates from a CSV file.

First, let's obtain the file:

In [225]:
import urllib.request
url = 'http://cocl.us/Geospatial_data'
filename = 'Geospatial_Coordinates.csv'
urllib.request.urlretrieve(url, filename)

('Geospatial_Coordinates.csv', <http.client.HTTPMessage at 0x16a5eafbfc8>)

Read the data into a dataframe:

In [226]:
df_coordinates = pd.read_csv('Geospatial_Coordinates.csv')
df_coordinates.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

Let's merge these dataframes based on postal code:

In [227]:
df_toronto = pd.merge(df_toronto, df_coordinates, on = 'Postal Code')
df_toronto.head()

Postal Code           Borough                                 Neighborhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494

In [228]:
df_toronto.shape

(103, 5)

## Part 3 - Explore and Cluster Toronto Neighborhoods

First, let us use geopy to find the coordinates for Toronto, ON:

In [229]:
!conda install -c conda-forge geopy --yes 
print("Geopy installed successfully!")

Solving environment: ...working... 
  - anaconda/win-64::openssl-1.1.1d-he774522_2
  - defaults/win-64::openssl-1.1.1d-he774522_2done

# All requested packages already installed.

Geopy installed successfully!


Import the necessary modules:

In [230]:
from geopy.geocoders import Nominatim
import folium

Determine coordinates for Toronto:

In [231]:
address = 'Toronto, ON'

geolocator = Nominatim(user_agent = 'to_explorer')
location = geolocator.geocode(address)
latitude_to = location.latitude
longitude_to = location.longitude
print('The geographical coordinates of Toronto are {}, {}.'.format(latitude_to, longitude_to))

The geographical coordinates of Toronto are 43.6534817, -79.3839347.


Now, let's ceate a map of Toronto with the Postal Codes superimposed on top:

In [232]:
# Create a map of Toronto using coordinates:
map_toronto = folium.Map(location = [latitude_to, longitude_to], zoom_start = 10)

# Add markers to our map:
for lat, lng, neighborhood, borough in zip(df_toronto['Latitude'], df_toronto['Longitude'], df_toronto['Neighborhood'], df_toronto['Borough']):
    label = '{} ({})'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html = True)
    folium.CircleMarker(
        [lat, lng],
        radius = 5,
        popup = label,
        color = 'blue',
        fill = True,
        fill_color = '#3186cc',
        fill_opacity = 0.7,
        parse_html = False).add_to(map_toronto)

map_toronto

Next, we will define our Foursquare credentials and version. Note that the contents of this cell are hidden for privacy:

In [233]:
# @hidden_cell
CLIENT_ID = 'UQIHUB4NIGKYAGZK0ILVKQC2OFTJFWA1T3JJ2322CWBZIDTO'
CLIENT_SECRET = 'LPTTUKXU0B5QFOKCPWUUWTJ33TKYDVFL0AC3TKSAQ3O1GLMN'
VERSION = '20180605'

### Explore Neighborhoods in Toronto

We need a function that will obtain the venues that are within a certain distance of a neighborhood:

In [234]:
def getNearbyVenues(names, latitudes, longitudes, radius=500, LIMIT = 100):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

Now we will create a dataframe of venues in Toronto for each neighborhood:

In [235]:
toronto_venues = getNearbyVenues(names = df_toronto['Neighborhood'],
                                latitudes = df_toronto['Latitude'],
                                longitudes = df_toronto['Longitude']
                                )

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Queen's Park, Ontario Provincial Government
Islington Avenue, Humber Valley Village
Malvern, Rouge
Don Mills
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
East Toronto, Broadview North (Old East York)
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmo

Let's look at the resulting dataframe:

In [236]:
print(toronto_venues.shape)
toronto_venues.head()

(2145, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0         Parkwoods              43.753259              -79.329656   
1         Parkwoods              43.753259              -79.329656   
2  Victoria Village              43.725882              -79.315572   
3  Victoria Village              43.725882              -79.315572   
4  Victoria Village              43.725882              -79.315572   

                    Venue  Venue Latitude  Venue Longitude  \
0         Brookbanks Park       43.751976       -79.332140   
1           Variety Store       43.751974       -79.333114   
2  Victoria Village Arena       43.723481       -79.315635   
3             Tim Hortons       43.725517       -79.313103   
4               Portugril       43.725819       -79.312785   

          Venue Category  
0                   Park  
1      Food & Drink Shop  
2           Hockey Arena  
3            Coffee Shop  
4  Portuguese Restaurant

How many venues were returned for each neighborhood?

In [237]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Agincourt                                                               4   
Alderwood, Long Branch                                                 11   
Bathurst Manor, Wilson Heights, Downsview North                        21   
Bayview Village                                                         4   
Bedford Park, Lawrence Manor East                                      23   
Berczy Park                                                            58   
Birch Cliff, Cliffside West                                             4   
Brockton, Parkdale Village, Exhibition Place                           23   
Business reply mail Processing Centre, South Ce...                     17   
CN Tower, King and Spadina, Railway Lands, Harb...                     14   
Caledonia-Fairbanks                                                     4   
Canada Post Gateway Processing Centre                                  12   
Cedarbrae                                                               8   
Central Bay Street                                                     63   
Christie                                                               17   
Church and Wellesley                                                   78   
Clarks Corners, Tam O'Shanter, Sullivan                                13   
Cliffside, Cliffcrest, Scarborough Village West                         3   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                             34   
Davisville North                                                        7   
Del Ray, Mount Dennis, Keelsdale and Silverthorn                        4   
Don Mills                                                              27   
Dorset Park, Wexford Heights, Scarborough Town ...                      6   
Downsview                                                              18   
Dufferin, Dovercourt Village                                           16   
East Toronto, Broadview North (Old East York)                           3   
Eringate, Bloordale Gardens, Old Burnhamthorpe,...                      7   
Fairview, Henry Farm, Oriole                                           65   
First Canadian Place, Underground city                                100   
Forest Hill North & West, Forest Hill Road Park                         4   
Garden District, Ryerson                                              100   
Glencairn                                                               4   
Golden Mile, Clairlea, Oakridge                                        10   
Guildwood, Morningside, West Hill                                       8   
Harbourfront East, Union Station, Toronto Islands                     100   
High Park, The Junction South                                          24   
Hillcrest Village                                                       5   
Humber Summit                                                           1   
Humberlea, Emery                                                        2   
Humewood-Cedarvale                                                      4   
India Bazaar, The Beaches West                                         22   
Kennedy Park, Ionview, East Birchmount Park                             7   
Kensington Market, Chinatown, Grange Park                              58   
Kingsview Village, St. Phillips, Martin Grove G...                      4   
Lawrence Manor, Lawrence Heights                                       13   
Lawrence Park                                                           3   
Leaside                                                                33   
Little Portugal, Trinity                                               44   
Malvern, Rouge                                                          1   
Milliken, Agincourt North, Steeles East, L'Amor.

How many unique categories can be curated from all the returned venues?

In [238]:
print("There are {} unique categories for venues.".format(len(toronto_venues['Venue Category'].unique())))

There are 272 unique categories for venues.


### Analyze Each Neighborhood

In [239]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
cols = list(toronto_onehot)
cols.insert(0, cols.pop(cols.index('Neighborhood')))
toronto_onehot = toronto_onehot.ix[:, cols]

toronto_onehot.head()

C:\Users\scott\Anaconda3\lib\site-packages\ipykernel_launcher.py:10: FutureWarning: 
.ix is deprecated. Please use
.loc for label based indexing or
.iloc for positional indexing

See the documentation here:
http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#ix-indexer-is-deprecated
  # Remove the CWD from sys.path while we load stuff.
C:\Users\scott\Anaconda3\lib\site-packages\pandas\core\indexing.py:822: FutureWarning: 
.ix is deprecated. Please use
.loc for label based indexing or
.iloc for positional indexing

See the documentation here:
http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#ix-indexer-is-deprecated
  retval = getattr(retval, self.name)._getitem_axis(key, axis=i)


Neighborhood  Accessories Store  Afghan Restaurant  Airport  \
0         Parkwoods                  0                  0        0   
1         Parkwoods                  0                  0        0   
2  Victoria Village                  0                  0        0   
3  Victoria Village                  0                  0        0   
4  Victoria Village                  0                  0        0   

   Airport Food Court  Airport Lounge  Airport Service  Airport Terminal  \
0                   0               0                0                 0   
1                   0               0                0                 0   
2                   0               0                0                 0   
3                   0               0                0                 0   
4                   0               0                0                 0   

   American Restaurant  Antique Shop  Aquarium  Art Gallery  Art Museum  \
0                    0             0         0            0           0   
1                    0             0         0            0           0   
2                    0             0         0            0           0   
3                    0             0         0            0           0   
4                    0             0         0            0           0   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  Auto Garage  \
0                    0                 0                   0            0   
1                    0                 0                   0            0   
2                    0                 0                   0            0   
3                    0                 0                   0            0   
4                    0                 0                   0            0   

   Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  \
0              0          0           0           0       0     0    0   
1              0          0           0           0       0     0    0   
2              0          0           0           0       0     0    0   
3              0          0           0           0       0     0    0   
4              0          0           0           0       0     0    0   

   Baseball Field  Baseball Stadium  Basketball Stadium  Beach  \
0               0                 0                   0      0   
1               0                 0                   0      0   
2               0                 0                   0      0   
3               0                 0                   0      0   
4               0                 0                   0      0   

   Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant  Bike Shop  \
0                0         0           0                   0          0   
1                0         0           0                   0          0   
2                0         0           0                   0          0   
3                0         0           0                   0          0   
4                0         0           0                   0          0   

   Bistro  Board Shop  Boat or Ferry  Bookstore  Boutique  \
0       0           0              0          0         0   
1       0           0              0          0         0   
2       0           0              0          0         0   
3       0           0              0          0         0   
4       0           0              0          0         0   

   Brazilian Restaurant  Breakfast Spot  Brewery  Bridal Shop  \
0                     0               0        0            0   
1                     0               0        0            0   
2                     0               0        0            0   
3                     0               0        0            0   
4                     0               0        0            0   

   Bubble Tea Shop  Building  Burger Joint  Burrito Place  Bus Line  \
0                0         0             0              0         0   
1                0    

How large is this dataframe?

In [240]:
toronto_onehot.shape

(2145, 272)

Now let's group the rows by neighborhood and determine the mean of the frequency of each venue category:

In [241]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Accessories Store  \
0                                           Agincourt           0.000000   
1                              Alderwood, Long Branch           0.000000   
2     Bathurst Manor, Wilson Heights, Downsview North           0.000000   
3                                     Bayview Village           0.000000   
4                   Bedford Park, Lawrence Manor East           0.000000   
5                                         Berczy Park           0.000000   
6                         Birch Cliff, Cliffside West           0.000000   
7        Brockton, Parkdale Village, Exhibition Place           0.000000   
8   Business reply mail Processing Centre, South C...           0.000000   
9   CN Tower, King and Spadina, Railway Lands, Har...           0.000000   
10                                Caledonia-Fairbanks           0.000000   
11              Canada Post Gateway Processing Centre           0.000000   
12                                          Cedarbrae           0.000000   
13                                 Central Bay Street           0.000000   
14                                           Christie           0.000000   
15                               Church and Wellesley           0.000000   
16            Clarks Corners, Tam O'Shanter, Sullivan           0.000000   
17    Cliffside, Cliffcrest, Scarborough Village West           0.000000   
18                     Commerce Court, Victoria Hotel           0.000000   
19                                         Davisville           0.000000   
20                                   Davisville North           0.000000   
21   Del Ray, Mount Dennis, Keelsdale and Silverthorn           0.000000   
22                                          Don Mills           0.000000   
23  Dorset Park, Wexford Heights, Scarborough Town...           0.000000   
24                                          Downsview           0.000000   
25                       Dufferin, Dovercourt Village           0.000000   
26      East Toronto, Broadview North (Old East York)           0.000000   
27  Eringate, Bloordale Gardens, Old Burnhamthorpe...           0.000000   
28                       Fairview, Henry Farm, Oriole           0.000000   
29             First Canadian Place, Underground city           0.000000   
30    Forest Hill North & West, Forest Hill Road Park           0.000000   
31                           Garden District, Ryerson           0.000000   
32                                          Glencairn           0.000000   
33                    Golden Mile, Clairlea, Oakridge           0.000000   
34                  Guildwood, Morningside, West Hill           0.000000   
35  Harbourfront East, Union Station, Toronto Islands           0.000000   
36                      High Park, The Junction South           0.000000   
37                                  Hillcrest Village           0.000000   
38                                      Humber Summit           0.000000   
39                                   Humberlea, Emery           0.000000   
40                                 Humewood-Cedarvale           0.000000   
41                     India Bazaar, The Beaches West           0.000000   
42        Kennedy Park, Ionview, East Birchmount Park           0.000000   
43          Kensington Market, Chinatown, Grange Park           0.000000   
44  Kingsview Village, St. Phillips, Martin Grove ...           0.000000   
45                   Lawrence Manor, Lawrence Heights           0.076923   
46                                      Lawrence Park           0.000000   
47                                            Leaside           0.000000   
48                           Little Portugal, Trinity           0.000000   
49                                     Malvern, Rouge           0.000000   
50  Milliken, Agincourt North, Steeles East, L'Amo...           0.000000   
51  Mimico NW, The Queensway West, South of Bloor,...           0.000000   
52           

The new size of this dataframe:

In [242]:
toronto_grouped.shape

(97, 272)

Let's print each neighborhood along with the top 5 most common venues:

In [243]:
num_top_venues = 5

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Agincourt----
                       venue  freq
0                     Lounge  0.25
1             Breakfast Spot  0.25
2  Latin American Restaurant  0.25
3               Skating Rink  0.25
4          Accessories Store  0.00


----Alderwood, Long Branch----
                venue  freq
0         Pizza Place  0.18
1            Pharmacy  0.09
2  Athletics & Sports  0.09
3      Sandwich Place  0.09
4                 Gym  0.09


----Bathurst Manor, Wilson Heights, Downsview North----
            venue  freq
0            Bank  0.10
1     Coffee Shop  0.10
2  Ice Cream Shop  0.05
3     Bridal Shop  0.05
4   Shopping Mall  0.05


----Bayview Village----
                 venue  freq
0                 Café  0.25
1                 Bank  0.25
2  Japanese Restaurant  0.25
3   Chinese Restaurant  0.25
4    Accessories Store  0.00


----Bedford Park, Lawrence Manor East----
                venue  freq
0    Sushi Restaurant  0.09
1  Italian Restaurant  0.09
2          Restaurant  0.09
3      Sandwi

                venue  freq
0         Coffee Shop  0.10
1                Café  0.04
2  Seafood Restaurant  0.03
3        Cocktail Bar  0.03
4  Italian Restaurant  0.03


----Studio District----
                 venue  freq
0                 Café  0.10
1          Coffee Shop  0.07
2              Brewery  0.05
3  American Restaurant  0.05
4               Bakery  0.05


----Summerhill West, Rathnelly, South Hill, Forest Hill SE, Deer Park----
                venue  freq
0                 Pub  0.12
1         Coffee Shop  0.12
2                Bank  0.06
3  Light Rail Station  0.06
4         Supermarket  0.06


----The Annex, North Midtown, Yorkville----
                           venue  freq
0                 Sandwich Place  0.13
1                           Café  0.13
2                    Coffee Shop  0.09
3  Vegetarian / Vegan Restaurant  0.04
4                     Donut Shop  0.04


----The Beaches----
               venue  freq
0                Pub  0.17
1               Park  0.17
2  He

Now we can put this into a pandas dataframe:

In [244]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [245]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Agincourt                Lounge   
1                           Alderwood, Long Branch           Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North                  Bank   
3                                  Bayview Village                  Café   
4                Bedford Park, Lawrence Manor East           Coffee Shop   

       2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0  Latin American Restaurant        Breakfast Spot          Skating Rink   
1               Dance Studio                   Gym           Coffee Shop   
2                Coffee Shop         Shopping Mall        Sandwich Place   
3         Chinese Restaurant                  Bank   Japanese Restaurant   
4             Sandwich Place    Italian Restaurant            Restaurant   

  5th Most Common Venue        6th Most Common Venue 7th Most Common Venue  \
0            Donut Shop                        Diner        Discount Store   
1          Skating Rink                     Pharmacy                   Pub   
2           Supermarket    Middle Eastern Restaurant      Sushi Restaurant   
3     Electronics Store  Eastern European Restaurant             Drugstore   
4      Sushi Restaurant                  Pizza Place             Juice Bar   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0   Distribution Center               Dog Run       Doner Restaurant  
1        Sandwich Place    Athletics & Sports                   Pool  
2            Restaurant         Deli / Bodega    Fried Chicken Joint  
3            Donut Shop      Doner Restaurant       Department Store  
4               Butcher     Indian Restaurant                   Café

### Cluster Toronto Neighborhoods

Here, we will run k-means to cluster the nighborhoods into 5 clusters.

In [272]:
# Import necessary modules:
from sklearn.cluster import KMeans

# Set number of clusters:
k = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# Run k-means clustering:
kmeans = KMeans(n_clusters = k, random_state = 0).fit(toronto_grouped_clustering)

# Check cluster labels generated for each row in the dataframe:
kmeans.labels_[:]

array([2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 0, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 0, 2, 2, 2, 0, 2, 2, 2, 2, 2, 2, 2, 3, 1, 0, 2, 2, 2,
       0, 2, 0, 2, 2, 2, 0, 2, 2, 2, 0, 2, 2, 2, 1, 2, 2, 0, 2, 2, 2, 0,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 0, 2, 2, 2, 2, 4, 2,
       0, 2, 2, 2, 2, 2, 2, 0, 2])

Create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood:

In [247]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = df_toronto

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head() # check the last columns!

Postal Code           Borough                                 Neighborhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  Cluster Labels  1st Most Common Venue  \
0  43.753259 -79.329656             0.0                   Park   
1  43.725882 -79.315572             2.0  Portuguese Restaurant   
2  43.654260 -79.360636             2.0            Coffee Shop   
3  43.718518 -79.464763             2.0         Clothing Store   
4  43.662301 -79.389494             2.0            Coffee Shop   

    2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0       Food & Drink Shop           Yoga Studio               Dog Run   
1            Hockey Arena          Intersection           Coffee Shop   
2                     Pub                Bakery                  Park   
3  Furniture / Home Store     Accessories Store           Coffee Shop   
4             Yoga Studio                  Bank              Beer Bar   

        5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                Dessert Shop    Dim Sum Restaurant                 Diner   
1  Financial or Legal Service    Dim Sum Restaurant                 Diner   
2                     Theater        Breakfast Spot                  Café   
3          Miscellaneous Shop           Event Space   Arts & Crafts Store   
4               Smoothie Shop        Sandwich Place          Burger Joint   

  8th Most Common Venue  9th Most Common Venue 10th Most Common Venue  
0        Discount Store    Distribution Center       Doner Restaurant  
1        Discount Store    Distribution Center                Dog Run  
2        Farmers Market             Restaurant  Performing Arts Venue  
3              Boutique  Vietnamese Restaurant              Gift Shop  
4         Burrito Place             Restaurant                   Café

In [252]:
toronto_merged['Cluster Labels']

0      0.0
1      2.0
2      2.0
3      2.0
4      2.0
5      NaN
6      2.0
7      2.0
8      2.0
9      2.0
10     2.0
11     4.0
12     2.0
13     2.0
14     2.0
15     2.0
16     0.0
17     2.0
18     2.0
19     2.0
20     2.0
21     0.0
22     2.0
23     2.0
24     2.0
25     2.0
26     2.0
27     2.0
28     2.0
29     2.0
30     2.0
31     2.0
32     2.0
33     2.0
34     2.0
35     0.0
36     2.0
37     2.0
38     2.0
39     2.0
40     2.0
41     2.0
42     2.0
43     2.0
44     2.0
45     2.0
46     2.0
47     2.0
48     2.0
49     0.0
50     3.0
51     2.0
52     2.0
53     2.0
54     2.0
55     2.0
56     2.0
57     1.0
58     2.0
59     2.0
60     2.0
61     0.0
62     2.0
63     2.0
64     0.0
65     2.0
66     0.0
67     2.0
68     0.0
69     2.0
70     2.0
71     2.0
72     2.0
73     2.0
74     2.0
75     2.0
76     2.0
77     0.0
78     2.0
79     2.0
80     2.0
81     2.0
82     2.0
83     2.0
84     2.0
85     0.0
86     2.0
87     2.0
88     2.0
89     2.0
90     2.0

We have two instances of NaN, which arose from a lack of venues in that neighborhood. We will have to deal with this by omitting them.

Now visualize the clusters on our map:

In [264]:
import matplotlib.cm as cm
import matplotlib.colors as colors
import math

# create map
map_clusters = folium.Map(location=[latitude_to, longitude_to], zoom_start=11)

# set color scheme for the clusters
x = np.arange(k)
ys = [i + x + (i*x)**2 for i in range(k)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    if math.isnan(cluster) == False:
        label = folium.Popup(str(poi) + ' Cluster ' + str(int(cluster)), parse_html=True)
        folium.CircleMarker(
            [lat, lon],
            radius=5,
            popup=label,
            color=rainbow[int(cluster)-1],
            fill=True,
            fill_color=rainbow[int(cluster)-1],
            fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

### Examining the Clusters

Cluster 0:

In [273]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0         North York             0.0                  Park   
16              York             0.0                  Park   
21              York             0.0                  Park   
35         East York             0.0                  Park   
49        North York             0.0                  Park   
61   Central Toronto             0.0                  Park   
64              York             0.0                  Park   
66        North York             0.0                  Park   
68   Central Toronto             0.0                  Park   
77         Etobicoke             0.0                  Park   
85       Scarborough             0.0                  Park   
91  Downtown Toronto             0.0                  Park   
98         Etobicoke             0.0                  Park   

   2nd Most Common Venue       3rd Most Common Venue 4th Most Common Venue  \
0      Food & Drink Shop                 Yoga Studio               Dog Run   
16                 Field                Hockey Arena                 Trail   
21                  Pool               Women's Store               Dog Run   
35           Coffee Shop           Convenience Store           Yoga Studio   
49                Bakery  Construction & Landscaping           Yoga Studio   
61           Swim School                    Bus Line           Yoga Studio   
64     Convenience Store                 Yoga Studio      Doner Restaurant   
66                  Bank           Convenience Store           Yoga Studio   
68                 Trail            Sushi Restaurant         Jewelry Store   
77           Pizza Place              Sandwich Place              Bus Line   
85            Playground                 Yoga Studio               Dog Run   
91                 Trail                  Playground      Department Store   
98                 River                 Yoga Studio   Distribution Center   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0           Dessert Shop    Dim Sum Restaurant                 Diner   
16           Yoga Studio               Dog Run          Dessert Shop   
21      Department Store          Dessert Shop    Dim Sum Restaurant   
35      Doner Restaurant    Dim Sum Restaurant                 Diner   
49      Doner Restaurant    Dim Sum Restaurant                 Diner   
61   Distribution Center          Dessert Shop    Dim Sum Restaurant   
64          Dessert Shop    Dim Sum Restaurant                 Diner   
66      Doner Restaurant    Dim Sum Restaurant                 Diner   
68          Dessert Shop    Dim Sum Restaurant                 Diner   
77        Discount Store      Department Store          Dessert Shop   
85      Department Store          Dessert Shop    Dim Sum Restaurant   
91          Dessert Shop    Dim Sum Restaurant                 Diner   
98      Department Store          Dessert Shop    Dim Sum Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0         Discount Store   Distribution Center       Doner Restaurant  
16    Dim Sum Restaurant                 Diner         Discount Store  
21                 Diner        Discount Store    Distribution Center  
35        Discount Store   Distribution Center                Dog Run  
49        Discount Store   Distribution Center                Dog Run  
61                 Diner        Discount Store                Dog Run  
64        Discount Store   Distribution Center                Dog Run  
66        Discount Store   Distribution Center                Dog Run  
68        Discount Store   Distribution Center                Dog Run  
77    Dim Sum Restaurant                 Diner    Distribution Center  
85                 Diner        Discount Store    Distribution Center  
91        Discount Store   Distribution Center                Dog Run  
98                 Diner        Discount Store                Dog Run

Cluster 1:

In [ ]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Cluster 2:

Cluster 3:

Cluster 4: